autogen/agentchat/contrib/capabilities/agentoptimizer.py

```diff
@@ -0,0 +1,439 @@
+1. Function name 
+2. Function description 
+3. JSON schema of arguments encoded as a string
+4. A list of package names imported by the function packages 
+5. The code implementation
+
+Below are the signatures of the current functions:
+List A: {current_signitures}. 
+The following list are the function signatures that you have after taking {actions_num} actions to manupulate List A:
+List B: {incumbent_signitures}. 
+
+{accumerated_experience}
+
+Here are {conversations_num} conversation histories of solving {conversations_num} tasks using List B.
+History:
+{conversations_history}
+
+{statistic_informations}
+
+According to the information I provide, please take one of four actions to manipulate list B using the functions you know. 
+Instead of returning TERMINATE directly or taking no action, you should try your best to optimize the function list. Only take no action if you really think the current list is optimal, as more actions will harm performance in future tasks. 
+Even adding a general function that can substitute the assistant’s repeated suggestions of Python code with the same functionality could also be helpful.
+"""
+
+
+class AgentOptimizer:
+    """
+    Base class for optimizing AutoGen agents. Specifically, it is used to optimize the function list of the userproxy-assistant agents pair.
+    More information could be found in the following paper: https://arxiv.org/abs/2402.11359
+    """
+
+    def __init__(self, max_actions_per_step: int, azure_config: dict):
+        """
+        (These APIs are experimental and may change in the future.)
+        Args: 
+            max_actions_per_step (int): the maximum number of actions that the optimizer can take in one step.
+            azure_config (dict): the configuration of Azure OpenAI API. It has the following filds:
+                - "AZURE_OPENAI_API_KEY" (str): the API key of Azure OpenAI.
+                - "api_version" (str): the version of the API.
+                - "azure_endpoint" (str): the endpoint of Azure OpenAI.
+                - "model" (str): the model of Azure OpenAI, and we suggest using "gpt-4-1106-preview".
+        """
+        self.max_actions_per_step = max_actions_per_step
+        self._max_trials = 3
+
+        self.registered_assistant = None
+        self.registered_user_proxy = None
+        self._conversations_history = []
+        self._conversations_performance = []
+        self._experience = []
+
+        self._pre_signatures = None
+        self._cur_signatures = None
+        
+        self.model = azure_config["model"]
+        self._oai_config = azure_config
+        os.environ["AZURE_OPENAI_API_KEY"] = azure_config["AZURE_OPENAI_API_KEY"] 
+        self._client = AzureOpenAI(
+            api_version=azure_config["api_version"],
+            azure_endpoint=azure_config["azure_endpoint"],
+        )
+
+    def register_agent(self, assistant: ConversableAgent, user_proxy: ConversableAgent):
```

Comment:
This API looks very restricted: only considering the userProxy-assistant case. Can we have a more general way to specify the agents?